File: problem_3/prime_factors.py

def primes():
    yield 2
    D = {}
    q = 3
    while True:
        p = D.pop(q, 0)
        if p:
            x = p + q
            while x in D: x += p
            D[x] = p
        else:
            yield q
            D[q*q] = 2*q
        q += 2

stop = 600851475143
for p in primes():
    if stop % p is 0:
        print p
    if stop < p:
        break

File: problem_3/prime_factors.hs
primes = 2 : primes'
  where
    primes' = sieve [3,5..] 9 primes'
    sieve (x:xs) q ps@ ~(p:t)
      | x < q = x : sieve xs q ps
      | True  = sieve [x | x <- xs, rem x p /= 0] (head t^2) t

main = print $ last $ [x | x <- takeWhile (< 10000) primes, 600851475143 `mod` x == 0]
